File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        q = deque()

        q.append((0, root))
        curr_level = 0
        new_list = []
        res = []
        while q:
            level, node = q.popleft()

            if level == curr_level:
                new_list.append(node.val)
            else:
                res.append(new_list[:])
                new_list = [node.val]

            curr_level = level
            if node.left:
                q.append((level+1, node.left))
            if node.right:
                q.append((level+1, node.right))

        res.append(new_list)
        return res

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        std::vector<int> rights;      
        auto dfs = [&](auto&&self, TreeNode* node, int level = 0)  -> void{
            if (node == nullptr)
                return;
            if(rights.size() == level)
            {
                rights.push_back(node->val);
            }

            self(self, node->right, level +1);
            self(self, node->left, level+1);
        };

        dfs(dfs, root);
        return rights;
    }
};
